File: client/src/test/java/nl/altindag/client/service/DispatchRebootServiceShould.java
package nl.altindag.client.service;

import dispatch.Http;
import nl.altindag.client.model.ClientResponse;
import org.asynchttpclient.*;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.ArgumentCaptor;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnitRunner;

import java.util.concurrent.CompletableFuture;

import static nl.altindag.client.ClientType.DISPATCH_REBOOT_HTTP_CLIENT;
import static nl.altindag.client.TestConstants.*;
import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.*;

@RunWith(MockitoJUnitRunner.class)
public class DispatchRebootServiceShould {

    @InjectMocks
    private DispatchRebootService victim;
    @Mock
    private Http httpClient;

    @Test
    @SuppressWarnings("unchecked")
    public void executeRequest() throws Exception {
        Response response = mock(Response.class);
        DefaultAsyncHttpClient defaultAsyncHttpClient = mock(DefaultAsyncHttpClient.class);
        ListenableFuture<Response> listenableFuture = mock(ListenableFuture.class);

        when(httpClient.client()).thenReturn(defaultAsyncHttpClient);
        when(defaultAsyncHttpClient.executeRequest(any(RequestBuilder.class))).thenReturn(listenableFuture);
        when(listenableFuture.toCompletableFuture()).thenReturn(CompletableFuture.completedFuture(response));

        when(response.getStatusCode()).thenReturn(200);
        when(response.getResponseBody()).thenReturn("Hello");

        ArgumentCaptor<RequestBuilder> requestBuilderArgumentCaptor = ArgumentCaptor.forClass(RequestBuilder.class);

        ClientResponse clientResponse = victim.executeRequest(HTTP_URL);

        assertThat(clientResponse.getStatusCode()).isEqualTo(200);
        assertThat(clientResponse.getResponseBody()).isEqualTo("Hello");

        verify(defaultAsyncHttpClient, times(1)).executeRequest(requestBuilderArgumentCaptor.capture());
        Request request = requestBuilderArgumentCaptor.getValue().build();
        assertThat(request.getUrl()).isEqualTo(HTTP_URL);
        assertThat(request.getMethod()).isEqualTo(GET_METHOD);
        assertThat(request.getHeaders().get(HEADER_KEY_CLIENT_TYPE)).isEqualTo(DISPATCH_REBOOT_HTTP_CLIENT.getValue());
    }

}

File: client/src/test/java/nl/altindag/client/service/DispatchRebootServiceShould.scala
/*
 * Copyright 2018 Thunderberry.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package nl.altindag.client.service

import dispatch.{Http, Req, Res}
import nl.altindag.client.TestConstants.HTTP_URL
import nl.altindag.client.util.SSLFactoryTestHelper
import org.assertj.core.api.Assertions.assertThat
import org.mockito.scalatest.MockitoSugar
import org.scalatest.funspec.AnyFunSpec

import scala.concurrent.{ExecutionContext, Future}
import scala.util.Try

class DispatchRebootServiceShould extends AnyFunSpec with MockitoSugar {

  describe("execute request") {
    val client = mock[Http]
    val response = mock[Res]
    implicit val executionContext: ExecutionContext = any[ExecutionContext]

    when(client(any[Req])).thenReturn(Future.fromTry(Try(response)))
    when(response.getResponseBody).thenReturn("Hello")
    when(response.getStatusCode).thenReturn(200)

    val victim = new DispatchRebootService(client)

    val clientResponse = victim.executeRequest(HTTP_URL)

    assertThat(clientResponse.getStatusCode).isEqualTo(200)
    assertThat(clientResponse.getResponseBody).isEqualTo("Hello")
  }

  describe("create dispatch reboot http client with ssl") {
    val sslFactory = SSLFactoryTestHelper.createSSLFactory(true, true)

    val client = new DispatchRebootHttpClientConfig()
      .createDispatchRebootHttpClient(sslFactory)

    assertThat(client).isNotNull
    verify(sslFactory, times(1)).getKeyManager
    verify(sslFactory, times(1)).getTrustManager
    verify(sslFactory, times(1)).getProtocols
    verify(sslFactory, times(1)).getCiphers
  }

}
